File: android/app/src/main/kotlin/app/filehop/filehop/FileHopNativePlugin.kt
package app.filehop.filehop

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Native command/event boundary.
 * Mission 07 adds LAN NSD browse only. No Wi-Fi Direct/Aware, no Mission 10 sockets.
 * Native adapters never own PeerIdentity/trust policy.
 *
 * Threading invariant (D-07-16): Flutter platform-channel method calls
 * arrive on the Android main thread, which is exactly the serialized
 * state authority owned by FileHopLanDiscovery. `startBrowse()` /
 * `stopBrowse()` verify that invariant explicitly.
 */
class FileHopNativePlugin(
    context: Context,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private val identityStore = FileHopIdentitySecretStore(context)
    private val lanDiscovery =
        FileHopLanDiscovery(context) { payload -> emitEvent(payload) }
    private val mainHandler = Handler(Looper.getMainLooper())

    companion object {
        const val BRIDGE_VERSION = 1
        const val COMMANDS = "app.filehop.native.commands"
        const val EVENTS = "app.filehop.native.events"

        /**
         * Engine detach is one-way, but NSD stop confirmation is asynchronous.
         * Keep a bounded retry window so an onStopDiscoveryFailed callback can
         * transition the discovery owner to UNCERTAIN and a later main-thread
         * retry can request the authoritative stop again.
         */
        private const val DETACH_STOP_RETRY_COUNT = 12
        private const val DETACH_STOP_RETRY_INTERVAL_MS = 500L

        fun registerWith(engine: FlutterEngine, context: Context): FileHopNativePlugin {
            val plugin = FileHopNativePlugin(context.applicationContext)
            MethodChannel(engine.dartExecutor.binaryMessenger, COMMANDS)
                .setMethodCallHandler(plugin)
            EventChannel(engine.dartExecutor.binaryMessenger, EVENTS)
                .setStreamHandler(plugin)
            return plugin
        }
    }

    private var eventSink: EventChannel.EventSink? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments
        if (args != null && args !is Map<*, *>) {
            result.error("invalidArgument", "payload must be a map", null)
            return
        }
        val map = (args as? Map<*, *>) ?: emptyMap<Any, Any>()
        val version = map["bridgeVersion"]
        if (version != null && version != BRIDGE_VERSION) {
            result.error(
                "invalidArgument",
                "incompatible bridgeVersion $version",
                null,
            )
            return
        }

        when (call.method) {
            "ping" ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "ok" to true,
                    ),
                )
            "observeAvailability" -> {
                val kind = map["kind"] as? String ?: "unknown"
                if (kind == "lan") {
                    result.success(lanDiscovery.availability())
                } else {
                    result.success(
                        mapOf(
                            "bridgeVersion" to BRIDGE_VERSION,
                            "kind" to kind,
                            "status" to "UNSUPPORTED",
                            "detail" to "Mission 02 skeleton: transport not implemented",
                        ),
                    )
                }
            }
            "startDiscovery" -> {
                val kind = map["kind"] as? String ?: "unknown"
                if (kind != "lan") {
                    result.error(
                        "unsupported",
                        "Mission 02 skeleton: startDiscovery is not implemented",
                        null,
                    )
                    return
                }
                val failure = lanDiscovery.startBrowse()
                if (failure != null) {
                    result.error("nativeFailure", failure, null)
                } else {
                    result.success(
                        mapOf(
                            "bridgeVersion" to BRIDGE_VERSION,
                            "ok" to true,
                        ),
                    )
                }
            }
            "stopDiscovery" -> {
                val kind = map["kind"] as? String ?: "unknown"
                if (kind != "lan") {
                    result.error(
                        "unsupported",
                        "Mission 02 skeleton: stopDiscovery is not implemented",
                        null,
                    )
                    return
                }
                lanDiscovery.stopBrowse()
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "ok" to true,
                    ),
                )
            }
            "connect",
            "disconnect",
            "openEndpoint",
            ->
                result.error(
                    "unsupported",
                    "Mission 10 owns LAN connect/openEndpoint; ${call.method} is not implemented",
                    null,
                )
            "identitySecret.store" -> handleIdentityStore(map, result)
            "identitySecret.load" -> handleIdentityLoad(map, result)
            "identitySecret.delete" -> handleIdentityDelete(map, result)
            "identitySecret.status" -> handleIdentityStatus(map, result)
            "identitySecret.hasAny" -> handleIdentityHasAny(result)
            "identitySecret.deleteAll" -> handleIdentityDeleteAll(result)
            else -> result.notImplemented()
        }
    }

    private fun handleIdentityStore(map: Map<*, *>, result: MethodChannel.Result) {
        val bytes = asByteArray(map["privateKeyBytes"])
        if (bytes == null) {
            result.error("invalidArgument", "privateKeyBytes must be exactly 32 bytes", null)
            return
        }
        when (val outcome = identityStore.store(bytes)) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            is String ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "reference" to outcome,
                    ),
                )
            else -> result.error("nativeFailure", "failed to protect identity secret", null)
        }
    }

    private fun handleIdentityLoad(map: Map<*, *>, result: MethodChannel.Result) {
        val reference = map["reference"] as? String
        if (reference.isNullOrEmpty()) {
            result.error("invalidArgument", "reference must be a non-empty string", null)
            return
        }
        when (val outcome = identityStore.load(reference)) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            is ByteArray ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "privateKeyBytes" to outcome,
                    ),
                )
            else -> result.error("nativeFailure", "failed to load identity secret", null)
        }
    }

    private fun handleIdentityDelete(map: Map<*, *>, result: MethodChannel.Result) {
        val reference = map["reference"] as? String
        if (reference.isNullOrEmpty()) {
            result.error("invalidArgument", "reference must be a non-empty string", null)
            return
        }
        when (val outcome = identityStore.delete(reference)) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            else ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "ok" to true,
                    ),
                )
        }
    }

    private fun handleIdentityStatus(map: Map<*, *>, result: MethodChannel.Result) {
        val reference = map["reference"] as? String
        if (reference.isNullOrEmpty()) {
            result.error("invalidArgument", "reference must be a non-empty string", null)
            return
        }
        when (val outcome = identityStore.status(reference)) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            is String ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "status" to outcome,
                    ),
                )
            else -> result.error("nativeFailure", "failed to query identity secret", null)
        }
    }

    private fun handleIdentityHasAny(result: MethodChannel.Result) {
        when (val outcome = identityStore.hasAny()) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            is Boolean ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "any" to outcome,
                    ),
                )
            else -> result.error("nativeFailure", "failed to query identity secrets", null)
        }
    }

    private fun handleIdentityDeleteAll(result: MethodChannel.Result) {
        when (val outcome = identityStore.deleteAll()) {
            is FileHopIdentitySecretStore.Failure ->
                result.error(outcome.code, outcome.message, null)
            else ->
                result.success(
                    mapOf(
                        "bridgeVersion" to BRIDGE_VERSION,
                        "ok" to true,
                    ),
                )
        }
    }

    private fun asByteArray(value: Any?): ByteArray? {
        return when (value) {
            is ByteArray -> value
            is List<*> -> {
                if (value.size != FileHopIdentitySecretStore.PRIVATE_KEY_LEN) {
                    return null
                }
                val out = ByteArray(value.size)
                for (index in value.indices) {
                    val number = value[index] as? Number ?: return null
                    out[index] = number.toByte()
                }
                out
            }
            else -> null
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun emitEvent(payload: Map<String, Any?>) {
        mainHandler.post {
            eventSink?.success(payload)
        }
    }

    fun detach() {
        // First stop is immediate. Follow-up calls are harmless while STOPPING
        // or STOPPED and become meaningful only if NSD reports an asynchronous
        // stop failure and FileHopLanDiscovery moves to UNCERTAIN.
        lanDiscovery.detach()
        for (attempt in 1..DETACH_STOP_RETRY_COUNT) {
            mainHandler.postDelayed(
                { lanDiscovery.detach() },
                attempt * DETACH_STOP_RETRY_INTERVAL_MS,
            )
        }
        eventSink = null
    }
}

File: ios/Runner/FileHopNativePlugin.swift
import Flutter
import UIKit

/// Native command/event boundary.
/// Mission 07 adds LAN Bonjour browse only. No Wi-Fi Aware, no Mission 10 sockets.
/// Native adapters never own PeerIdentity/trust policy.
final class FileHopNativePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  static let bridgeVersion = 1
  static let commands = "app.filehop.native.commands"
  static let events = "app.filehop.native.events"

  private var eventSink: FlutterEventSink?
  private lazy var lanDiscovery = FileHopLanDiscovery { [weak self] payload in
    self?.emitEvent(payload)
  }

  static func register(with registrar: FlutterPluginRegistrar) {
    let instance = FileHopNativePlugin()
    let commandChannel = FlutterMethodChannel(
      name: Self.commands,
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: commandChannel)
    let eventChannel = FlutterEventChannel(
      name: Self.events,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(instance)
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let map = call.arguments as? [String: Any] ?? [:]
    if let version = map["bridgeVersion"] as? Int, version != Self.bridgeVersion {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "incompatible bridgeVersion \(version)",
          details: nil
        )
      )
      return
    }
    if call.arguments != nil && (call.arguments as? [String: Any]) == nil {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "payload must be a map",
          details: nil
        )
      )
      return
    }

    switch call.method {
    case "ping":
      result(["bridgeVersion": Self.bridgeVersion, "ok": true])
    case "observeAvailability":
      let kind = map["kind"] as? String ?? "unknown"
      if kind == "lan" {
        result(lanDiscovery.availability())
      } else {
        result([
          "bridgeVersion": Self.bridgeVersion,
          "kind": kind,
          "status": "UNSUPPORTED",
          "detail": "Mission 02 skeleton: transport not implemented",
        ])
      }
    case "startDiscovery":
      let kind = map["kind"] as? String ?? "unknown"
      if kind != "lan" {
        result(
          FlutterError(
            code: "unsupported",
            message: "Mission 02 skeleton: \(call.method) is not implemented",
            details: nil
          )
        )
        return
      }
      if let error = lanDiscovery.startBrowse() {
        result(FlutterError(code: "nativeFailure", message: error, details: nil))
      } else {
        result(["bridgeVersion": Self.bridgeVersion, "ok": true])
      }
    case "stopDiscovery":
      let kind = map["kind"] as? String ?? "unknown"
      if kind != "lan" {
        result(
          FlutterError(
            code: "unsupported",
            message: "Mission 02 skeleton: \(call.method) is not implemented",
            details: nil
          )
        )
        return
      }
      lanDiscovery.stopBrowse()
      result(["bridgeVersion": Self.bridgeVersion, "ok": true])
    case "connect", "disconnect", "openEndpoint":
      result(
        FlutterError(
          code: "unsupported",
          message: "Mission 10 owns LAN connect/openEndpoint; \(call.method) is not implemented",
          details: nil
        )
      )
    case "identitySecret.store":
      handleIdentityStore(map: map, result: result)
    case "identitySecret.load":
      handleIdentityLoad(map: map, result: result)
    case "identitySecret.delete":
      handleIdentityDelete(map: map, result: result)
    case "identitySecret.status":
      handleIdentityStatus(map: map, result: result)
    case "identitySecret.hasAny":
      replyIdentity(FileHopIdentitySecretStore.hasAny(), result: result, anyKey: true)
    case "identitySecret.deleteAll":
      replyIdentity(FileHopIdentitySecretStore.deleteAll(), result: result, ok: true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleIdentityStore(map: [String: Any], result: FlutterResult) {
    guard let data = privateKeyData(map["privateKeyBytes"]) else {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "privateKeyBytes must be exactly 32 bytes",
          details: nil
        )
      )
      return
    }
    switch FileHopIdentitySecretStore.store(privateKey: data) {
    case let failure as FileHopIdentitySecretStore.Failure:
      result(FlutterError(code: failure.code, message: failure.message, details: nil))
    case let reference as String:
      result(["bridgeVersion": Self.bridgeVersion, "reference": reference])
    default:
      result(
        FlutterError(
          code: "nativeFailure",
          message: "failed to protect identity secret",
          details: nil
        )
      )
    }
  }

  private func handleIdentityLoad(map: [String: Any], result: FlutterResult) {
    guard let reference = map["reference"] as? String, !reference.isEmpty else {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "reference must be a non-empty string",
          details: nil
        )
      )
      return
    }
    switch FileHopIdentitySecretStore.load(reference: reference) {
    case let failure as FileHopIdentitySecretStore.Failure:
      result(FlutterError(code: failure.code, message: failure.message, details: nil))
    case let bytes as FlutterStandardTypedData:
      result(["bridgeVersion": Self.bridgeVersion, "privateKeyBytes": bytes])
    default:
      result(
        FlutterError(
          code: "nativeFailure",
          message: "failed to load identity secret",
          details: nil
        )
      )
    }
  }

  private func handleIdentityDelete(map: [String: Any], result: FlutterResult) {
    guard let reference = map["reference"] as? String, !reference.isEmpty else {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "reference must be a non-empty string",
          details: nil
        )
      )
      return
    }
    replyIdentity(FileHopIdentitySecretStore.delete(reference: reference), result: result, ok: true)
  }

  private func handleIdentityStatus(map: [String: Any], result: FlutterResult) {
    guard let reference = map["reference"] as? String, !reference.isEmpty else {
      result(
        FlutterError(
          code: "invalidArgument",
          message: "reference must be a non-empty string",
          details: nil
        )
      )
      return
    }
    switch FileHopIdentitySecretStore.status(reference: reference) {
    case let failure as FileHopIdentitySecretStore.Failure:
      result(FlutterError(code: failure.code, message: failure.message, details: nil))
    case let status as String:
      result(["bridgeVersion": Self.bridgeVersion, "status": status])
    default:
      result(
        FlutterError(
          code: "nativeFailure",
          message: "failed to query identity secret",
          details: nil
        )
      )
    }
  }

  private func replyIdentity(
    _ outcome: Any,
    result: FlutterResult,
    ok: Bool = false,
    anyKey: Bool = false
  ) {
    switch outcome {
    case let failure as FileHopIdentitySecretStore.Failure:
      result(FlutterError(code: failure.code, message: failure.message, details: nil))
    case let any as Bool where anyKey:
      result(["bridgeVersion": Self.bridgeVersion, "any": any])
    case is Bool:
      result(["bridgeVersion": Self.bridgeVersion, "ok": true])
    default:
      result(
        FlutterError(
          code: "nativeFailure",
          message: "identity secret operation failed",
          details: nil
        )
      )
    }
  }

  private func privateKeyData(_ value: Any?) -> Data? {
    if let typed = value as? FlutterStandardTypedData,
      typed.data.count == FileHopIdentitySecretStore.privateKeyLength
    {
      return typed.data
    }
    if let data = value as? Data, data.count == FileHopIdentitySecretStore.privateKeyLength {
      return data
    }
    return nil
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  /// Discovery callbacks arrive on FileHopLanDiscovery's private queue.
  /// Flutter event sinks are main-thread objects, and cancellation may race a
  /// queued emission, so resolve the current sink only after hopping to main.
  private func emitEvent(_ payload: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      self?.eventSink?(payload)
    }
  }

  func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    lanDiscovery.detach()
    eventSink = nil
  }
}
